main.js

```diff
@@ -210,24 +210,42 @@ async function main() {
             materials[i].mapIndex = -1;
         }
     }
-    let materialInfo = [];
-    for (let i = 0; i < materials.length; i++) {
-        materialInfo.push({
-            metalness: materials[i].metalness,
-            roughness: materials[i].roughness,
-            emissive: new THREE.Vector3(materials[i].emissive.r, materials[i].emissive.g, materials[i].emissive.b),
-            mapIndex: materials[i].mapIndex,
-            color: new THREE.Vector3(materials[i].color.r, materials[i].color.g, materials[i].color.b)
-        });
-    }
-    while (materialInfo.length < 64) {
-        materialInfo.push({
-            metalness: 0.0,
-            roughness: 0.0,
-            mapIndex: -1,
-            color: new THREE.Vector3(1.0, 1.0, 1.0),
-            emissive: new THREE.Vector3(0.0, 0.0, 0.0)
-        });
+    const materialInfo1 = [];
+    const materialInfo2 = [];
+    const materialInfo3 = [];
+    const MAX_MATERIALS = 64
+    for (let i = 0; i < MAX_MATERIALS; i++) {
+        if (i >= materials.length) {
+            materialInfo1.push(
+                0.0, // metalness
+                0.0, // roughness
+                -1, // mapIndex
+                0,
+            );
+            materialInfo2.push(
+                1.0, 1.0, 1.0, // color
+                0,
+            );
+            materialInfo3.push(
+                0.0, 0.0, 0.0, // emissive
+                0,
+            );
+        } else {
+            materialInfo1.push(
+                materials[i].metalness,
+                materials[i].roughness,
+                materials[i].mapIndex,
+                0, // padding
+            );
+            materialInfo2.push(
+                materials[i].emissive.r, materials[i].emissive.g, materials[i].emissive.b,
+                0, // padding
+            );
+            materialInfo3.push(
+                materials[i].color.r, materials[i].color.g, materials[i].color.b,
+                0, // padding
+            );
+        }
```

Comment:
This should be a single array, but I can't find an efficient way to pack it into multiple rows.

Should maybe create a vec3/float packer since colors here are [0-1].

Reply:
Yes a vec3/float packer could work - I already use one in the shader.